Add namespace and project backend persistence to container_projects

screeley44 opened the pull request:
adding namepaces from kubernetes and projects from openshift into container_projects table

screeley44 commented:
@simon3z - not sure this is best approach - on initial refresh (first save into the container_projects) - I would get duplicate records and couldn't find a good way to squash those, it's because both kubernetes and openshift were using the same label for the hash (i.e. container_projects) - I made the labels more descriptive so they wouldn't overlap (i.e. container_projects_k8 and container_projects_openshift) - which resulted in two additional keys in the save_inventory - maybe you know a cleaner way to do this?

Fryguy commented:
Is this related to #2673 ?

screeley44 commented:
@Fryguy - I believe so, the decision was made to use a single entity (i.e. container_projects) to hold the similar information from both k8 and openshift (namespaces and projects respectively), at least that was my understanding... but as stated, not sure my approach is correct  @simon3z can probably elaborate on the design decision a bit more.

blomquisg commented:
@screeley44, sounds like this is wip?

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -61,6 +62,10 @@ def get_endpoints(inventory)
     end
 
+    def get_namespaces(inventory)
+      process_collection(inventory["namespace"], :container_projects_k8) { |n| parse_namespaces(n) }
```
Use `:container_projects` here.

simon3z commented on vmdb/app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -2,7 +2,7 @@ module EmsRefresh::SaveInventoryContainer
     target = ems if target.nil?
     child_keys = [:container_nodes, :container_replicators, :container_groups,
-                  :container_services, :container_routes, :container_projects]
+                  :container_services, :container_routes, :container_projects_k8, :container_projects_openshift]
```
You can just leave this as it was.

simon3z commented on vmdb/app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -12,7 +12,23 @@ def save_ems_container_inventory(ems, hashes, target = nil)
   end
 
-  def save_container_projects_inventory(ems, hashes, target = nil)
+  def save_container_projects_k8_inventory(ems, hashes, target = nil)
```
This shouldn't be necessary.

simon3z commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -14,16 +14,20 @@ def get_routes(inventory)
 
       def get_projects(inventory)
-        process_collection(inventory["project"], :container_projects) { |n| parse_project(n) }
+        process_collection(inventory["project"], :container_projects_openshift) { |n| parse_project(n) }
```
I think this should become simply:

    inventory["project"].each { |item| parse_project(item) }

Since this part won't be actually returning any new data, it would just augment the data already collected.

/cc @Fryguy

simon3z commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -14,16 +14,20 @@ def get_routes(inventory)
-            :labels       => parse_labels(project),
-            :display_name => project.displayName
-        )
+
```
This would probably become:

    project = @data_index.fetch_path(:container_projects, :by_name, project.name)

    unless project.metadata.annotations.nil?
        project.merge!(
            :display_name => project.metadata.annotations.displayName,
            :labels       => parse_labels(project)
        )
    end

Basically here you would fetch the already stored/indexed `namespace` and you would add the additional information you get from parsing `projects`.

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -61,6 +62,10 @@ def get_endpoints(inventory)
     end
 
+    def get_namespaces(inventory)
+      process_collection(inventory["namespace"], :container_projects) { |n| parse_namespaces(n) }
```
I thought I had commented about adding also (maybe I haven't submitted it):

    @data[:container_projects].each do |ns|
        @data_index.store_path(:container_projects, :by_name, ns[:name], ns)
    end

screeley44 commented:
@simon3z - still missing something...the project.merge! is producing expected results:

{:ems_ref=>"56ba01a9-07d2-11e5-9950-525400f10bba",
 :name=>"openshift",
 :creation_timestamp=>"2015-05-31T20:19:28Z",
 :resource_version=>"55",
 :display_name=>nil}
{:ems_ref=>"0b45c619-07d6-11e5-9950-525400f10bba",
 :name=>"test-openshift",
 :creation_timestamp=>"2015-05-31T20:45:59Z",
 :resource_version=>"126",
 :display_name=>"Hello Test OpenShift"}
{:ems_ref=>"562c547f-07d2-11e5-9950-525400f10bba",
 :name=>"default",
 :creation_timestamp=>"2015-05-31T20:19:27Z",
 :resource_version=>"57",
 :display_name=>nil}

however, the hash being passed into the save_container_inventory has duplicate entries...

[{:ems_ref=>"562c547f-07d2-11e5-9950-525400f10bba",
  :name=>"default",
  :creation_timestamp=>"2015-05-31T20:19:27Z",
  :resource_version=>"57",
  :display_name=>nil},
 {:ems_ref=>"56ba01a9-07d2-11e5-9950-525400f10bba",
  :name=>"openshift",
  :creation_timestamp=>"2015-05-31T20:19:28Z",
  :resource_version=>"55",
  :display_name=>nil},
 {:ems_ref=>"0b45c619-07d6-11e5-9950-525400f10bba",
  :name=>"test-openshift",
  :creation_timestamp=>"2015-05-31T20:45:59Z",
  :resource_version=>"126",
  :display_name=>"Hello Test OpenShift"},
 {:ems_ref=>"56ba01a9-07d2-11e5-9950-525400f10bba",
  :name=>"openshift",
  :creation_timestamp=>"2015-05-31T20:19:28Z",
  :resource_version=>"55",
  :display_name=>nil},
 {:ems_ref=>"0b45c619-07d6-11e5-9950-525400f10bba",
  :name=>"test-openshift",
  :creation_timestamp=>"2015-05-31T20:45:59Z",
  :resource_version=>"126",
  :display_name=>"Hello Test OpenShift"},
 {:ems_ref=>"562c547f-07d2-11e5-9950-525400f10bba",
  :name=>"default",
  :creation_timestamp=>"2015-05-31T20:19:27Z",
  :resource_version=>"57",
  :display_name=>nil}]
so as a result on the initial save (no records exist in container_projects), we have duplicates, on every subsequent refresh, the data looks correct.

simon3z commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -14,17 +14,17 @@ def get_routes(inventory)
 
       def get_projects(inventory)
-        process_collection(inventory["project"], :container_projects) { |n| parse_project(n) }
+        process_collection(inventory["project"].each, :container_projects) { |item| parse_project(item) }
```
You haven't followed my suggestion here of just using:

    inventory["project"].each { |item| parse_project(item) }

I think that's why you're collecting the items twice.

simon3z commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -14,17 +14,17 @@ def get_routes(inventory)
+            :display_name => project_item.metadata.annotations.displayName
+          )
+        end
+        project
```
This is redundant.

simon3z commented on vmdb/app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -23,8 +23,7 @@ def save_container_projects_inventory(ems, hashes, target = nil)
 
-    save_inventory_multi(:container_projects, ems, hashes, deletes, [:ems_ref],
-                         :labels)
+    save_inventory_multi(:container_projects, ems, hashes, deletes, [:ems_ref])
```
Is this change required?

simon3z commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -14,17 +14,16 @@ def get_routes(inventory)
 
       def get_projects(inventory)
-        process_collection(inventory["project"], :container_projects) { |n| parse_project(n) }
+        inventory["project"].each { |item| parse_project(item) }
```
cc @Fryguy this is correct from our point of view, I am not sure if that's ok with you.

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -217,6 +226,12 @@ def parse_endpoint(entity)
+    def parse_namespaces(container_projects)
+      new_result = parse_base_item(container_projects)
+      new_result.except!(:namespace)
+      new_result
```
Maybe we can have a one-liner?

    def parse_namespaces(container_projects)
      parse_base_item(container_projects).except(:namespace)
    end

simon3z commented:
@screeley44 is it possible to add spec tests in `kubernetes_refresher_spec.rb`?

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -17,6 +17,7 @@ def ems_inv_to_hashes(inventory)
       get_pods(inventory)
       get_endpoints(inventory)
       get_services(inventory)
+      get_namespaces(inventory)
```
Move this after:

    get_nodes(inventory)

Parsing namespaces is the first thing you'll need to create relationships.

screeley44 commented:
@simon3z - this is done with the exception of the spec test - not quite sure how to use those, but can you review the backend work - it is working correctly from my point of view

screeley44 commented:
@simon3z - might need some minor tweaks, but all is tested and functional - have backend working for container_projects (namespaces/projects) and relationships on the UI from both summary screen and left hand accordian.  Including a few screenshots to give a sample.
![3048_projects_summary_showing_relationships](https://cloud.githubusercontent.com/assets/6773375/8138818/54823616-111c-11e5-8225-684c29ce51e4.png)
![3408-service_to_project_relationship](https://cloud.githubusercontent.com/assets/6773375/8138817/54820182-111c-11e5-85e8-049fa8748a4c.png)

simon3z commented:
@screeley44 you're repeating `except!(:namespace)` several time in the code. It's worth considering to remove it once and for all from `parse_base_item`.

But more importantly. If you remove `namespace` (either from here or from `parse_base_item`) you're basically breaking all the relationships based on the `namespace` (see all the `by_namespace_and_name` entries).

screeley44 commented:
@simon3z - not sure how to avoid the unknown attribute error then?  namespace doesn't exist in the tables any more and if I pass it in the hash results, it complains about it.

last_refresh_error: "unknown attribute: namespace"

screeley44 commented:
@simon3z - and when I don't include the namespace, I didn't notice it broke anything in the relationships, like container_groups uses by_namespace_and_by_name and didn't seem to have any negative impact - let me know, and I'll experiment

simon3z commented:
@screeley44 I personally solved it this way:  https://github.com/simon3z/manageiq/commit/98be1a9b144e2765b0abac5acf6d1fc98c8cb537

I am not sure if it's optimal.

simon3z commented:
@screeley44 also please look at `vmdb/product/views/ContainerGroup.yaml` (the container group table cloumns) and consider to rename `namespace` to `project_name` (you also have to implement `def project_name` in `container_group.rb`

screeley44 commented:
@simon3z - np on the table view columns, but for the project_name, are you saying in the textual_summary for container groups we should have a field for project_name?  isn't that redundant since we have a relationship link to project that already displays the name?  or are you saying something else that i'm missing?

screeley44 commented:
![project_name_added_to_table](https://cloud.githubusercontent.com/assets/6773375/8309167/e93580d8-1996-11e5-8d66-3c7cce30dee6.png)

screeley44 commented:
@simon3z @h-kataria @Fryguy - I think this is ready to merge but failing on some travis tests - maybe the migration script is not correct - using the wrong method of "change" ?

screeley44 commented:
@simon3z @Fryguy @h-kataria - travis passes now, I just needed to add the type to my migration script - please review and merge at your convenience...thanks

screeley44 commented:
@blomquisg @Fryguy @simon3z - I keep getting notified that I need to rebase (guessing that is due to new merges going in), I did that twice now and all is mergeable again - I believe this is ready when you guys are, thanks

chessbyte commented on vmdb/app/controllers/mixins/containers_common_mixin.rb:
```diff
@@ -47,6 +47,50 @@ def show_container(record, controller_name, display_name)
+        @bottom_msg = "* You are not authorized to view "
+        + pluralize(@view.extras[:total_count] - @view.extras[:auth_count], "other #{title.singularize}")
+        + " on this " + ui_lookup(:tables => @table_name)
+      end
```
This looks like a lot of similar code.  Can it be refactored?  Are there any new tests/specs for this chunk of code?

cc @dclarizio @h-kataria @martinpovolny

chessbyte commented:
@screeley Can you add some specs?  If you need help, @tenderlove can assist you.

screeley44 commented:
@chessbyte @tenderlove @simon3z - I updated the kubernetes_refresher_spec.rb and that seems to work fine (adding in the container_project table), and I also added a controller spec as I saw container_replicator_controller_spec.rb existed, so modeled after it but it keeps failing, is this needed or is the kubernetes_refresher_spec.rb good enough?  Any thoughts on what I'm missing for it?

simon3z commented on vmdb/app/views/layouts/listnav/_container_service.html.haml:
```diff
@@ -30,3 +30,9 @@
+            = link_to("#{ui_lookup(:table => "container_project")}: #{@record.container_project.name}",
+                {:controller => "container_project", :action => 'show', :id => @record.container_project.id.to_s},
+                {:title => _("Show this container service's parent %s") % ui_lookup(:table => "container_project")},
+                '/images/icons/16/link_external.gif')
```
@screeley44 this last argument `'/images/icons/16/link_external.gif'` seems unneeded. Isn't it?

simon3z replied:
@screeley44 there are also other to be removed I believe (replicators, route, etc). Somehow related to #3267

tenderlove commented on vmdb/app/views/container_route/show.html.haml:
```diff
@@ -1 +1,4 @@
-= render :partial => @showtype
+- if %w(container_routes container_projects).include?(@display)
```
Where do these strings come from?  I'm really concerned about including a list of magic strings.

tenderlove commented on vmdb/app/views/layouts/listnav/_container_group.html.haml:
```diff
@@ -55,3 +55,9 @@
+            = link_to("#{ui_lookup(:table => "container_project")}: #{@record.container_project.name}",
+                {:controller => "container_project", :action => 'show', :id => @record.container_project.id.to_s},
+                {:title => _("Show this container groups's parent %s") % ui_lookup(:table => "container_project")},
+                '/images/icons/16/link_external.gif')
```
`link_to` only takes 3 parameters, so this can't work. 73fee40bce7e483d1b62be75b6fea96216cad4fc

tenderlove commented on vmdb/app/views/layouts/listnav/_container_replicator.html.haml:
```diff
@@ -20,3 +20,10 @@
+            = link_to("#{ui_lookup(:table => "container_project")}: #{@record.container_project.name}",
+                {:controller => "container_project", :action => 'show', :id => @record.container_project.id.to_s},
+                {:title => _("Show this container service's parent %s") % ui_lookup(:table => "container_project")},
+                '/images/icons/16/link_external.gif')
```
Same as above.

tenderlove commented on vmdb/app/views/layouts/listnav/_container_route.html.haml:
```diff
@@ -20,3 +20,10 @@
+            = link_to("#{ui_lookup(:table => "container_project")}: #{@record.container_project.name}",
+                {:controller => "container_project", :action => 'show', :id => @record.container_project.id.to_s},
+                {:title => _("Show this container service's parent %s") % ui_lookup(:table => "container_project")},
+                '/images/icons/16/link_external.gif')
```
Same as above.

tenderlove commented:
@screeley44 Can you try running the tests on your machine and checking the output?  It looks like if you just run `rspec ./spec/controllers/container_project_controller_spec.rb:15 `, then check the output of test.log, it will tell you what is breaking.  I suspect it's one of these `link_to` calls that have 4 parameters.

simon3z commented on vmdb/app/helpers/container_group_helper/textual_summary.rb:
```diff
@@ -4,13 +4,13 @@ module ContainerGroupHelper::TextualSummary
   def textual_group_relationships
     # Order of items should be from parent to child
-    items = %w(ems services containers container_node lives_on)
+    items = %w(ems services containers container_node container_project lives_on)
```
Please keep this ordered, `container_project` should go first.

simon3z replied:
Sorry, I meant first of the entities (`ems` excluded):

    ems container_project services containers container_node lives_on

simon3z commented on vmdb/app/helpers/container_project_helper/textual_summary.rb:
```diff
@@ -9,7 +9,7 @@ def textual_group_properties
 
   def textual_group_relationships
-    items = %w(ems)
+    items = %w(ems container_groups container_services container_routes container_replicators)
```
Order should be:

    ems container_routes container_services container_replicators container_groups

simon3z commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -29,7 +28,7 @@ def textual_group_port_configs
 
   def textual_group_relationships
-    items = %w(ems container_groups)
+    items = %w(ems container_groups container_project)
```
Order should be `ems container_project container_groups`

moolitayer commented:
as @screeley44 is on holiday I will be addressing the comments
made here at PR #3340 that contains the commit from this PR (under his authorship)

kbrock commented:
@screeley44 / @simon3z Is this PR still needed? Does #3340 cover it?

screeley44 commented:
@kbrock @simon3z @mkanoor - oops, my bad, I didn't notice Mooli's comment that he was going to finish this up, sorry.  So yeah, I think #3340 probably covers it, but will let @simon3z have final say on that.  Sorry, was bored for a few minutes here on my PTO and thought I would dig into this PR but will stop now :-)

simon3z commented:
@screeley44 I didn't want you to spoil your PTO with these last small fixes (it goes without saying that I'll make sure that your authorship of the patch will be maintained).

@kbrock #3340 is the most updated series of patch, you can review and merge that.

chessbyte commented:
@screeley44 Please rebase and push again.

simon3z commented:
@chessbyte this can be closed as we merged #3340
